ykllvmwrap/src/jitmodbuilder.cc

```diff
@@ -1470,38 +1501,72 @@ class JITModBuilder {
     for (size_t Idx = 0; Idx < InpTrace.Length(); Idx++) {
       // Update the previously executed BB in the most-recent frame (if it's
       // mappable).
-      optional<IRBlock> MaybeIB = InpTrace[Idx];
+      TraceLoc Loc = InpTrace[Idx];
 
-      // If we previously saw a call to foreign code, there will now be an
-      // "unmappable hole" in the trace.
-      if (ExpectUnmappable) {
-        assert(!MaybeIB.has_value());
-        // FIXME: Allow foreign code to call back to mappable code.
+      if (UnmappableRegion *UR = Loc.getUnmappableRegion()) {
+        // The trace entered a region of unmappable foreign code.
         //
-        // We've just seen an unmappable region in the trace. There are never
-        // two unmappable regions in a row, so the next block is guaranteed to
-        // be mappable. Since we don't yet allow foreign code to "call back"
-        // into mappable code, we assume that the subsequent mappable block is
-        // the result of the foreign code returning to mappable code.
-        assert(!CallStack.curFrame().getMappableFrame());
-        CallStack.popFrame();
-        ExpectUnmappable = false;
-        tryStopOutlining();
+        // There are never two unmappable regions in a row, so the next block
+        // (if there is one) is guaranteed to be mappable. The question is: will
+        // the foreign code be *returning* into mappable code, or will it be
+        // *calling back* deeper into mappable code? The trace decoder keeps
+        // track of the stack effects of foreign code (assuming its control flow
+        // isn't bonkers), so we are able to use that information to decide.
+
+        if (InpTrace.Length() == Idx + 1) {
+          // This unmappable region is the end of the trace, so there's no
+          // need to maintain the stack any more.
+          continue;
+        }
+
+        assert(!CallStack.curMappableFrame());
+        if (UR->StackAdjust < 0) {
+          // The stack got smaller as a result of executing the foreign code,
```

Comment:
How does this happen -- `longjmp`?

Reply:
Oh, I see, so mostly this comment really just refers to `return` (the only other way I *think* this could happen is `longjmp`)?

Reply:
The trace compiler already aborts traces containing longjump and setjmp. That should be enough for now, right? I'd expect anyone implementing that to write a test anyway.

Reply:
If there's a `longjmp` in unmappable code, the trace compiler won't see it?

Reply:
I hate it when you are right :)

Reply:
Bask in it :)

Fixed in 3eeb456022b425fb87ca8297b8baa88895d99faf

Reply:
Moi?